libc/dirent/readdir.c

```diff
@@ -0,0 +1,309 @@
+/*
+ * Phoenix-RTOS
+ *
+ * POSIX.1-2017 standard library functions tests
+ * HEADER:
+ *    - dirent.h
+ * TESTED:
+ *    - readdir()
+ * Copyright 2023 Phoenix Systems
+ * Author: Arkadiusz Kozlowski
+ *
+ * This file is part of Phoenix-RTOS.
+ *
+ * %LICENSE%
+ */
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <unistd.h>
+#include <dirent.h>
+#include <signal.h>
+
+#include <sys/types.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <string.h>
+
+#include <unity_fixture.h>
+
+#include "common.h"
+#include "dirent_helper_functions.h"
+
+#define MAIN_DIR            "test_readdir"
+#define INO_T_TEST_MAX_DIRS 10
+
+
+TEST_GROUP(dirent_readdir);
+
+TEST_SETUP(dirent_readdir)
+{
+	mkdir(MAIN_DIR, 0777);
+
+	mkdir(MAIN_DIR "/dir1", 0777);
+	mkdir(MAIN_DIR "/dir2", 0777);
+
+	mkdir(MAIN_DIR "/dir1/nest1", 0777);
+	mkdir(MAIN_DIR "/dir1/nest2", 0777);
+
+	mkdir(MAIN_DIR "/dir2/nest1", 0777);
+	mkdir(MAIN_DIR "/dir2/nest2", 0777);
+
+	FILE *files[] = {
+		fopen(MAIN_DIR "/file1.txt", "w+"),
+		fopen(MAIN_DIR "/file2.dat", "w+"),
+		fopen(MAIN_DIR "/file3.json", "w+")
+	};
+
+	if (files[0]) {
+		fprintf(files[0], "Some data");
+		fprintf(files[1], "Some other data");
+
+
+		fclose(files[0]);
+		fclose(files[1]);
+		fclose(files[2]);
+	}
+}
+
+
+TEST_TEAR_DOWN(dirent_readdir)
+{
+	rmdir(MAIN_DIR "/dir1/nest1");
+	rmdir(MAIN_DIR "/dir1/nest2");
+
+	rmdir(MAIN_DIR "/dir2/nest1");
+	rmdir(MAIN_DIR "/dir2/nest2");
+
+	rmdir(MAIN_DIR "/dir1");
+	rmdir(MAIN_DIR "/dir2");
+
+	remove(MAIN_DIR "/file1.txt");
+	remove(MAIN_DIR "/file2.dat");
+	remove(MAIN_DIR "/file3.json");
+
+	rmdir(MAIN_DIR);
+}
+
+
+TEST(dirent_readdir, long_name_directory_check)
+{
+	DIR *dp = test_opendir_asserted(MAIN_DIR);
+	struct dirent *info;
+	char longDirName[NAME_MAX];
+	char longDirPath[NAME_MAX + 2 + sizeof(MAIN_DIR)];
+
+	longDirName[NAME_MAX - 1] = '\0';
+
+	TEST_ASSERT_NOT_NULL(dp);
+
+	memset(longDirName, 'a', NAME_MAX - 1);
+	sprintf(longDirPath, MAIN_DIR "/%s", longDirName);
```

Comment:
`NAME_MAX` do not include terminating byte
```suggestion
	DIR *dp;
	struct dirent *info;
	char longDirName[NAME_MAX + 1];
	char longDirPath[NAME_MAX + 1 + sizeof(MAIN_DIR)];

        memset(longDirName, 'a', NAME_MAX);
	longDirName[NAME_MAX] = '\0';
	
        dp = opendir(MAIN_DIR);
	TEST_ASSERT_NOT_NULL(dp);

	sprintf(longDirPath, MAIN_DIR "/%s", longDirName);
```